tiled/adapters/hdf5.py

```diff
@@ -13,25 +20,303 @@
-    return ArrayAdapter.from_array(dataset, metadata=getattr(dataset, "attrs", {}))
+def ndslice_from_string(
+    arg: str,
+) -> Tuple[Union[int, builtins.slice, EllipsisType], ...]:
+    """Parse and convert a string representation of a slice
 
+    For example, '(1:3, 4, 1:5:2, ...)' is converted to (slice(1, 3), 4, slice(1, 5, 2), ...).
+    """
+    if not (arg.startswith("[") and arg.endswith("]")) and not (
+        arg.startswith("(") and arg.endswith(")")
+    ):
+        raise ValueError("Slice must be enclosed in square brackets or parentheses.")
+    result: list[Union[int, builtins.slice, EllipsisType]] = []
+    for part in arg[1:-1].split(","):
+        if part.strip() == ":":
+            result.append(builtins.slice(None))
+        elif m := re.match(r"^(\d+):$", part.strip()):
+            start, stop = int(m.group(1)), None
+            result.append(builtins.slice(start, stop))
+        elif m := re.match(r"^:(\d+)$", part.strip()):
+            start, stop = 0, int(m.group(1))
+            result.append(builtins.slice(start, stop))
+        elif m := re.match(r"^(\d+):(\d+):(\d+)$", part.strip()):
+            start, stop, step = map(int, m.groups())
+            result.append(builtins.slice(start, stop, step))
+        elif m := re.match(r"^(\d+):(\d+)$", part.strip()):
+            start, stop = map(int, m.groups())
+            result.append(builtins.slice(start, stop))
+        elif m := re.match(r"^(\d+)$", part.strip()):
+            result.append(int(m.group()))
+        elif part.strip() == "...":
+            result.append(Ellipsis)
+        else:
+            raise ValueError(f"Invalid slice part: {part}")
+        # TODO: cases like "::n" or ":4:"
+    return tuple(result)
+
+
+def parse_hdf5_tree(
+    tree: Union[h5py.File, h5py.Group, h5py.Dataset]
+) -> Union[dict[str, Union[Any, None]], None]:
+    """Parse an HDF5 file or group into a nested dictionary structure
+
+    the resulting tree structure represenets any groups as nested dictionaries ans datasets as None.
+
+    Parameters
+    ----------
+    tree : h5py.File or h5py.Group
+        The file or group to parse
+
+    Returns
+    -------
+    dict
+        A nested dictionary structure representing the HDF5 file or group
+    """
+    res: dict[str, Union[Any, None]] = {}
+
+    if isinstance(tree, h5py.Dataset):
+        return None
+
+    for key, val in tree.items():
+        res[key] = parse_hdf5_tree(val)
+
+    return res
+
+
+def get_hdf5_attrs(
+    file_uri: str,
+    dataset: Optional[str] = None,
+    swmr: bool = SWMR_DEFAULT,
+    libver: str = "latest",
+    **kwargs: Optional[Any],
+) -> JSON:
+    """Get attributes of an HDF5 dataset"""
+    file_path = path_from_uri(file_uri)
+    with h5py.File(file_path, "r", swmr=swmr, libver=libver, **kwargs) as _file:
+        node = _file[dataset] if dataset else _file
+        d = dict(getattr(node, "attrs", {}))
+        for k, v in list(d.items()):
```

Comment:
I think this `list` call would unnecessarily traverse this an extra time.

```suggestion
        for k, v in d.items():
```